pkg/csi/service/common/vsphereutil.go

```diff
@@ -1096,43 +1104,43 @@ func GetNodeVMsWithAccessToDatastore(ctx context.Context, vc *vsphere.VirtualCen
 	}
 
 	// Get datastore object.
-	dsInfoObj, err := getDatastoreInfoObj(ctx, vc, dsURL)
+	dsInfoObjList, err := getDatastoreInfoObjList(ctx, vc, dsURL)
 	if err != nil {
 		return nil, logger.LogNewErrorf(log, "failed to retrieve datastore object using datastore "+
 			"URL %q. Error: %+v", dsURL, err)
 	}
-	// Get datastore host mounts.
-	var ds mo.Datastore
-	err = dsInfoObj.Properties(ctx, dsInfoObj.Reference(), []string{"host"}, &ds)
-	if err != nil {
-		return nil, logger.LogNewErrorf(log, "failed to get host mounts from datastore %q. Error: %+v",
-			dsURL, err)
-	}
-
-	// For each host mount, get the list of VMs.
-	for _, host := range ds.Host {
-		hostObj := &vsphere.HostSystem{
-			HostSystem: object.NewHostSystem(vc.Client.Client, host.Key),
-		}
-		var hs mo.HostSystem
-		err = hostObj.Properties(ctx, hostObj.Reference(), []string{"vm"}, &hs)
+	for _, dsInfoObj := range dsInfoObjList {
+		// Get datastore host mounts.
+		var ds mo.Datastore
+		err = dsInfoObj.Properties(ctx, dsInfoObj.Reference(), []string{"host"}, &ds)
 		if err != nil {
-			return nil, logger.LogNewErrorf(log, "failed to retrieve virtual machines from host %q. Error: %+v",
-				hostObj.String(), err)
+			return nil, logger.LogNewErrorf(log, "failed to get host mounts from datastore %q. Error: %+v",
+				dsURL, err)
 		}
-		// For each VM on the host, check if the VM is a NodeVM participating in the k8s cluster.
-		for _, vm := range hs.Vm {
-			if _, exists := nodeMap[vm]; exists {
-				accessibleNodes = append(accessibleNodes, object.NewVirtualMachine(vc.Client.Client, vm))
+		// For each host mount, get the list of VMs.
+		for _, host := range ds.Host {
+			hostObj := &vsphere.HostSystem{
+				HostSystem: object.NewHostSystem(vc.Client.Client, host.Key),
 			}
-			// Check if the length of accessible nodes is equal to total count of
-			// NodeVMs in this cluster. If yes, we can stop searching for more VMs.
-			if len(accessibleNodes) == totalNodes {
-				log.Infof("Nodes that have access to datastore %q are %+v", dsURL, accessibleNodes)
-				return accessibleNodes, nil
+			var hs mo.HostSystem
+			err = hostObj.Properties(ctx, hostObj.Reference(), []string{"vm"}, &hs)
+			if err != nil {
+				return nil, logger.LogNewErrorf(log, "failed to retrieve virtual machines from host %q. Error: %+v",
+					hostObj.String(), err)
+			}
+			// For each VM on the host, check if the VM is a NodeVM participating in the k8s cluster.
+			for _, vm := range hs.Vm {
+				if _, exists := nodeMap[vm]; exists {
+					accessibleNodes = append(accessibleNodes, object.NewVirtualMachine(vc.Client.Client, vm))
+				}
 			}
 		}
-
+		// Check if the length of accessible nodes is equal to total count of
+		// NodeVMs in this cluster. If yes, we can stop searching for more VMs.
+		if len(accessibleNodes) == totalNodes {
+			log.Infof("Nodes that have access to datastore %q are %+v", dsURL, accessibleNodes)
+			return accessibleNodes, nil
+		}
```

Comment:
should this block be inside For loop => " for _, vm := range hs.Vm {" ?

Reply:
Yes. We can exit early by doing so.